components/remote_settings/src/lib.rs

```diff
@@ -0,0 +1,209 @@
+pub mod error;
+use config::RemoteSettingsConfig;
+pub use error::{RemoteSettingsError, Result};
+mod config;
+use rs_client::{Client, ClientConfig};
+use std::{fs::File, io::prelude::Write, sync::Mutex};
+
+uniffi_macros::include_scaffolding!("remotesettings");
+
+pub struct RemoteSettings {
+    pub config: RemoteSettingsConfig,
+    client: Mutex<Client>,
+}
+
+impl RemoteSettings {
+    pub fn new(config: RemoteSettingsConfig) -> Result<Self> {
+        Ok(RemoteSettings {
+            config: config.clone(),
+            client: Mutex::new(Client::new(config.into())?),
+        })
+    }
+
+    pub fn get_records(&self) -> Result<Vec<JsonObject>> {
+        let client = self.client.lock().unwrap();
+        let resp = client.get_records()?;
+        self.parse_records(resp.text().to_string())
```

Comment:
Wow yeah, that is very slick. To double-check my understanding, basically, the compiler is able to propagate the `serde_json::Value` type through the generic chain in `RemoteSettingsResponse` and the placeholder in `Ok(resp.json::<RemoteSettingsResponse<_>>()?.data)` will force the compiler to reify that generic during compile time?

Reply:
Yup exactly!!

the `resp.json` function takes a type parameter in 
https://github.com/mozilla/application-services/blob/9657aebb7450c5b58e8b9a88bec12bd5e9e0f700/components/viaduct/src/lib.rs#L246-L252

We specify that type parameter when we do `::<...>` (this is called a t[urbofish](https://github.com/jplatte/turbo.fish)!)

The type we provide is `RemoteSettingsResponse<T>`, which itself takes a type parameter. We do `RemoteSettingsResponse<_>` to let the compiler infer what `T` is, it can do that because the return type of `get_records` includes the `JsonOnbject`, so the compiler infers that we wanted `RemoteSettingsResponse<JsonObject>`.


I wrote that not knowing whether we'd expect `data` to have concrete types, buf if we expect `data` to always be an arbitrary schema, it might be worth removing the generic in `RemoteSettingsResponse` so it's always holding on to a `data: Vec<JsonObject>` which should function the same way!